README.md
# Armor PubSub

ArmorPubSub is a publish/subscribe library written in JavaScript. It can be used in both Browser or NodeJS. It needs no dependencies and can be brought in as an `UMD` module. It exposes a global variable named `ArmorPubSub` when it be loaded globally.

ArmorPubSub asynchronously publishes topics by default, you can also synchronously publish topics if you want, there is a so-called `invoke queue` that is used to sort all publish/unsubscribe to be called in order, so you don't worry about embedding publish/unsubscribe, they behave like asynchronous calls which actually are synchronous in an `invoke loop`, you can get a promise which would be resolved when an invoke loop is over, or be rejected if an error occurred.

# Installation

`npm install armor-pubsub`

# Usage

```js
import ArmorPubSub from "armor-pubsub";

// Assign ArmorPubSub to any object that will can publish/subscribe topic.
let foo = Object.assign({}, ArmorPubSub);

// Make a subscrption to a topic and return an unsubscribe handler which is a function
// that could be called whenever you want to unsubscribe this specific subscription.
let unsubscribe = foo.sub("topic", msg => console.log(msg));

// Publish a topic with a content
foo.pub("topic", "hello world");

// Return a promise which would be resolved when unsubscription has been done.
let promise = unsubscribe();
```

# Terms

## namespace

A seperated space to store the subscriptions of topics. The default namespace name is `default`.

```js
const foo = Object.assign({ id: "foo" }, ArmorPubSub);

foo.sub("hello", callback, { ns: foo.id });

foo.pub("hello", content, { ns: foo.id });

// wrap pub/sub into a specific namespace.
foo.subPrivate = function(topic, callback, options) {
  return this.sub(topic, callback, Object.assign({}, options, { ns: this.id }));
};

foo.pubPrivate = function(topic, content, options) {
  return this.pub(topic, content, Object.assgin({}, options, { ns: this.id }));
};
```

## invoke-queue

You can choose to publish/unsubscribe topic in a synchronous way or an asynchronous way.

ArmorPubSub uses `setTimeout` to asynchronously invoke callbacks of subscription, so it won't block the current call stack and publish/unsubscribe will be called one-by-one in turn.

The synchronously invoking callbacks is a little more complicated, the publish/unsubscribe might be embedded in another one, the callback might has been removed when iterating the callbacks. So there is a so-called `invoke-queue`, all of the invoke of the publish/unsubscibe would be pushed into this queue to wait for being executed in turn instead of calling them immediately. The duration between the first invoke pushed into the queue and the last one has been pushed out is called an `invoke-loop`. In an `invoke-loop`, every one of calling publish/unsubscribe would return a same promise that won't be resolved until the current `invoke-loop` has finished.

You can see the steps when using `invoke-queue` to deal with publish/unsubscribe in the testcase below:

```js
it("Embed publish/subscribe", done => {
  let step = 0; // counter to show the current step.
  let p1, p2, p3, p4;

  // subscribe to topic x.
  foo.sub("x", () => {
    // This line would be met in the first invoke-loop.
    // The publish won't be executed immediately, instead,
    // it would be pushed into invoke-queue and wait for being executed.
    // It returns a promise of the first invoke-loop.
    p2 = foo.pub("y", null, { sync: true });
    p2.then(() => {
      expect(++step).toBe(4);
    });
    // Publish another topic z,
    // it returns the same promise of the first invoke-loop.
    p4 = foo.pub("z");
    expect(p2).toBe(p4);

    expect(++step).toBe(1);
  });

  // subscribe to topic y and save the unsubscribe handler.
  let promiseToDestroy = foo.sub("y", () => {
    expect(++step).toBe(2);
  });

  // subscribe to topic z
  foo.sub("z", () => {
    // As topic z is published asynchronously, this callback would be called by using setTimeout
    expect(++step).toBe(7);
    done();
  });

  expect(p2).toBeUndefined(); // The value of p2 is still undefined.

  // No one has been called until we decide to make a synchronous publish.
  // It returns the promise of the first invoke-loop.
  p1 = foo.pub("x", void 0, { sync: true });
  p1.then(() => {
    expect(++step).toBe(5);
  });
  // Make sure all the promises returned is the same one in a single invoke-loop.
  expect(p1).toBe(p2);

  // The first invoke-loop is over.
  // We call the unsubscribe handler to start another invoke-loop.
  p3 = promiseToDestroy();
  p3.then(() => {
    expect(++step).toBe(6);
  });
  // Make sure the two promises are different.
  expect(p3 === p1).toBeFalsy();

  expect(++step).toBe(3);
});
```

# APIs

## pub(topic:string, [content:any], [options:object])

Publish a topic and return a promise that will be resolved when the `invoke-loop` is over.

- `topic`: This argument should be a string, if it given an `undefined` or `null`, it would be force to transform to an empty string.
- `content`: This optional argument would be passed into callback function provided when subscribing a topic as the first argument.
- `options`: This optional argument must be a plain object that can include the keys as below:
  - `ns:string`: The namespace of topic, if it given an `undefined` or `null`, it would be tranformed to default name `default`.
  - `sync:bool`: If given true, it will synchronously call callbacks for the topic, otherwise they will be called asynchronously. Default as `false`.

## sub(topic:string, callback:function, [options:object])

Make a subscription to the topic with a callback, return an `unsubscribe` function which can be called to unsubscribe this specific subscription.

A promise that will be resolved when the `invoke-loop` is over is returned when calling `unsubscribe` function.

- `topic`: This argument should be a string, if it given an `undefined` or `null`, it would be force to transform to an empty string.
- `callback`: The function bound to the topic will be called when the topic is published.
- `options`: This optional argument must be a plain object that can include the keys as below:
  - `ns:string`: The namespace of topic, if it given an `undefined` or `null`, it would be tranformed to default name `default`.
  - `ctx:object`: The context of `callback`, default value is the current subscriber.

## subOnce(topic:string, callback:function, [options:object])

It behaves like `sub`, but the subscription will be unsubscribed as soon as the topic has once been published.

## unsub([topic:string], [callback:function], [options])

Unsubscribe a specific topic or multiple topics filtered by `topic`, `callback`, `ns` and `ctx`.

The subscriptions of a subscriber will be unsubscribed if they meet the conditions as below:

- `topic`: If this is given `undefined` or `null`, it will match all topics, otherwise it must match the topic of subscription.
- `callback`: If this is given `undefined` or `null`, it will match all callbacks, otherwise it must match the callback of subscription.
- `ns`: If this is given `undefined` or `null`, it will match all namespaces, otherwise it must match the ns of subscription.
- `ctx`: if this is given falsy, it will match all ctxs, otherwise it must match the ns of subscription.

## hasSub([topic:string], [callback:function], [options])

To find out if a subscription has been made by the subscriber, the signature of `hasSub` is same as `unsub`, but it returns `true` if the first subscription is found, or `false` if no subscription is found.


CHANGELOG.md
# v0.2.0

Supported on exeuting all publish/unsubscribe in order, and returning a promise that would be resolved when all of the tasks in a queue have been executed.

# v0.1.0

Finished the first version of armor-pubsub.


spec/pubsubSpec.js
/*
 * @Author: Xavier Yin 
 * @Date: 2018-08-10 14:52:44 
 * @Last Modified by: Xavier Yin
 * @Last Modified time: 2018-08-13 15:42:14
 */
import PubSub from "armor-pubsub";

describe("Armor-pubsub test cases.", () => {
  let foo, bar;

  beforeEach(() => {
    foo = Object.assign({ id: "foo" }, PubSub);
    bar = Object.assign({ id: "bar" }, PubSub);
  });

  afterEach(() => {
    foo.unsub();
    bar.unsub();
  });

  it("[sub] Subscribe a topic", done => {
    let cancel = foo.sub("x", null);
    expect(cancel).toBeUndefined(); // callback 不是函数不能订阅。

    let m = 200;
    cancel = foo.sub("x", function(x) {
      expect(x).toBe(m); // x 传参来自 pub 发布
      expect(this).toBe(foo); // 默认上下文为当前订阅者
      done(); // 异步调用
    });
    foo.pub("x", m);
    cancel();
    let hasAnySubscription = foo.hasSub("x");
    expect(hasAnySubscription).toBe(false); // 通过 sub() 返回的 cancel 函数撤销订阅。
  });

  it("[pub] Synchronously publish a topic", () => {
    let m = 0;
    foo.sub("x", function(x) {
      m += x;
    });
    foo.pub("x", 10, { sync: true });
    expect(m).toBe(10); // 因为是同步发布，所以此时 m 值为加运算后的值。
  });

  it("[pub] Asynchronously publish a topic", done => {
    let m = 0;
    foo.sub("x", function(x) {
      m += x;
      if (m === 10) done();
    });
    foo.pub("x", 10);
    expect(m).toBe(0); // 因为是异步发布，所以此时 m 还未进行加法运算。
  });

  it("[pub] Subscribe once", () => {
    let m = 0;
    foo.subOnce("x", function(x) {
      m += x;
    });
    foo.pub("x", 10, { sync: true });
    foo.pub("x", 10, { sync: true });
    expect(m).toBe(10); // 因为发布一次后订阅即被解除，所以此时 m 值为10。
  });

  it("[pub] Publish and subscribe with namespace", () => {
    let m = 0;
    let ns = "foo";
    foo.sub(
      "x",
      function(x) {
        m += x;
      },
      { ns }
    );
    foo.pub("x", 10, { sync: true });
    expect(m).toBe(0); // 因为命名空间不对，所以订阅回调未执行
    foo.pub("x", 10, { sync: true, ns });
    expect(m).toBe(10); // 命名空间正确，所以此时 m 值为10。
  });

  it("[pub] Unsubscribe", () => {
    let [t1, c1, n1, ctx1] = ["x", () => {}, "x", {}];
    let [t2, c2, n2, ctx2] = ["y", () => {}, "y", {}];

    foo.sub(t2, c2, { ns: n2, ctx: ctx2 });

    let sub = () => {
      foo.sub(t1, c1, { ns: n1, ctx: ctx1 });
    };
    sub();
    let has1 = () => foo.hasSub(t1);
    let has2 = () => foo.hasSub(t2);

    expect(has1()).toBe(true); // 证明订阅了
    expect(has2()).toBe(true); // 证明订阅了

    foo.unsub(t1);
    expect(has1()).toBe(false); // 证明解绑了

    sub();
    foo.unsub(null, c1);
    expect(has1()).toBe(false); // 证明解绑了

    sub();
    foo.unsub(null, null, { ns: n1 });
    expect(has1()).toBe(false); // 证明解绑了

    sub();
    foo.unsub(null, null, { ctx: ctx1 });
    expect(has1()).toBe(false); // 证明解绑了

    sub();
    foo.unsub(t1, c1, { ctx: ctx1, ns: n1 });
    expect(has1()).toBe(false); // 证明解绑了

    expect(has2()).toBe(true); // 证明没有解绑不应该解绑的订阅

    sub();
    foo.unsub();
    expect(has1()).toBe(false); // 证明所有的都解绑了
    expect(has2()).toBe(false);
  });

  it("[pub] Synchronous publishing and subscribing should be called in queue", () => {
    let m = 0;
    foo.sub("x", () => {
      foo.pub("y", null, { sync: true }); // #1
      foo.unsub("z"); // #2
      expect(m).toBe(0); // 此时 pub('y') 被压入队列尚未执行，因此 m 为 0；
    });
    foo.sub("y", () => {
      m += 10;
      foo.pub("z", null, { sync: true }); // #3
    });
    foo.sub("z", () => {
      m += 100;
    });
    foo.pub("x", null, { sync: true }); // #4

    // 因此 #4 执行时，#1，#2 都压入队列并未立即调用。
    // #3 执行时，z 发布压入队列未立即执行。
    // 因此整个过程队列中的操作顺序为：
    // 1. 发布 x
    // 2. 发布 y
    // 4. 取消订阅 z
    // 5. 发布 z
    // 所以到执行第5个任务时，z 已经被取消订阅，但整个过程都是同步操作，所以此时 m 为 10
    expect(m).toBe(10);
  });

  it("[pub] Asynchronous publishing and subscribing should be called in queue", done => {
    let m = 0;
    foo.sub("x", () => {
      foo.pub("y"); // #1
      foo.unsub("z"); // #2
      expect(m).toBe(0); // 此时 pub('y') 被异步压入队列尚未执行，因此 m 为 0；
    });

    foo.sub("y", () => {
      m += 10;
      foo.pub("z"); // #3
      expect(m).toBe(10);
    });

    foo.sub("z", () => {
      m += 100;
      expect(m).toBe(100000); // 永远不会到达这里
    });

    foo.pub("x"); // #4

    // 因为整个过程都是异步将任务推入队列执行，因此整个过程队列中都始终只有一个任务，它们分别是：
    // 1. 发布 x
    // 2. 发布 y
    // 4. 取消订阅 z
    // 5. 发布 z
    // 所以第5次清空队列时，z 已经被取消订阅，所以此时 m 为 10
    expect(m).toBe(0);

    setTimeout(() => {
      expect(m).toBe(10); // 证明最终值为 10
      done();
    }, 1000);
  });

  it("[promise] Get promise when embedding publish/subscribe", done => {
    let m = 0; // 计步器
    let p1, p2, p3, p4; // Promises

    // 订阅主题 x
    foo.sub("x", () => {
      // 在第一次队列循环中，发布主题 y，
      // 本次发布操作被压入本次队列循环，等待执行。
      // 获取本次队列循环的完结 PromiseA
      p2 = foo.pub("y", null, { sync: true });
      p2.then(() => {
        expect(++m).toBe(4);
      });
      p4 = foo.pub("z");
      expect(p2).toBe(p4);
      expect(++m).toBe(1);
    });

    // 订阅主题 y，同时保留取消订阅函数
    let promiseToDestroy = foo.sub("y", () => {
      expect(++m).toBe(2);
    });

    // 订阅主题 z
    foo.sub("z", () => {
      // 因为 z 是异步发布，本回调函数处于 macro-task 事件循环中
      // 所以本回调函数是最后被调用。
      expect(++m).toBe(7);
      done();
    });

    expect(p2).toBeUndefined(); // 此时 p2 仍然是 undefined
    // 发布主题 x，得到本次队列完结 PromiseA。
    p1 = foo.pub("x", void 0, { sync: true });
    p1.then(() => {
      expect(++m).toBe(5);
    });
    // 此时 p2 已被赋值 PromiseA
    expect(p2 === void 0).toBeFalsy();
    expect(p1).toBe(p2);

    // 取消订阅主题 y，得到本次队列完结 PromiseB
    p3 = promiseToDestroy();
    p3.then(() => {
      expect(++m).toBe(6);
    });
    expect(p3 === p1).toBeFalsy();

    // 同步执行的代码
    // 此之前的 expect 判断，只有同步发布的x, y 主题
    expect(++m).toBe(3);
  });
});


src/index.js
/*
 * @Author: Xavier Yin 
 * @Date: 2018-08-10 10:58:35 
 * @Last Modified by: Xavier Yin
 * @Last Modified time: 2018-08-10 16:24:02
 */
import subscribe from "./subscribe";
import unsubscribe, { unsubscribeByIdInQueue } from "./unsubscribe";
import publish from "./publish";
import { isFunction, isEmptyArray } from "./utils";

const pubsub = {
  /**
   * 判断是否订阅了某个 Topic，以及使用 callback, ctx, ns 进行筛选
   * @param {*} topic
   * @param {*} callback
   * @param {*} options
   */
  hasSub(topic, callback, options) {
    let pubsub = this._armorPubSub;
    if (isEmptyArray(pubsub)) return false;
    let { ns, ctx } = options || {};
    let subscribing, i;
    for (i = 0; i < pubsub.length; i++) {
      subscribing = pubsub[i];
      let { topic: t, callback: cb, ns: n, ctx: c } = subscribing;
      if (
        (topic == void 0 || topic === t) &&
        (callback == void 0 || callback === cb) &&
        (ns == void 0 || ns === n) &&
        (!ctx || ctx === c)
      ) {
        return true;
      }
    }
    return false;
  },

  sub(topic, callback, options) {
    if (!isFunction(callback)) return;
    let { ns, ctx } = options || {};
    let id = subscribe(this, topic, callback, { ns, ctx });
    return function() {
      unsubscribeByIdInQueue(id);
    };
  },

  subOnce(topic, callback, options) {
    if (!isFunction(callback)) return;
    let { ns, ctx } = options || {};
    let id = subscribe(this, topic, callback, { ns, ctx, once: true });
    return function() {
      unsubscribeByIdInQueue(id);
    };
  },

  pub(topic, content, options) {
    publish(topic, content, options);
    return this;
  },

  unsub(topic, callback, options) {
    unsubscribe(this, topic, callback, options);
    return this;
  }
};

module.exports = pubsub;


src/publish.js
/*
 * @Author: Xavier Yin 
 * @Date: 2018-08-09 15:13:42 
 * @Last Modified by: Xavier Yin
 * @Last Modified time: 2018-08-13 16:12:11
 */

import { safeNs, safeTopic, isEmptyArray, queue } from "./utils";
import { getRegistry } from "./registry";
import { unsubscribeById } from "./unsubscribe";

function pubSync(subscribing, content) {
  let { callback, ctx } = subscribing;
  callback.call(ctx, content);
}

function pubAsync(subscribing, content) {
  setTimeout(function() {
    pubSync(subscribing, content);
  });
}

function publishApi(topic, content, options) {
  let { ns, sync } = options || {};
  ns = safeNs(ns);
  topic = safeTopic(topic);

  let registry = getRegistry(ns);
  if (!registry) return;

  // 没有注册表
  let subscribings = registry[topic];
  // 没有订阅者
  if (isEmptyArray(subscribings)) return;

  let onces = [];
  let func = sync ? pubSync : pubAsync;
  let index = 0;
  let subscribing;
  while (index < subscribings.length) {
    subscribing = subscribings[index++];
    func(subscribing, content);
    let { once } = subscribing;
    if (once) onces.push(subscribing);
  }

  for (let i = 0; i < onces.length; i++) {
    unsubscribeById(onces[i]["id"]);
  }
}

export default function publish(topic, content, options) {
  return queue({ fn: publishApi, args: [topic, content, options] });
}


src/registry.js
/*
 * @Author: Xavier Yin 
 * @Date: 2018-08-09 15:11:21 
 * @Last Modified by: Xavier Yin
 * @Last Modified time: 2018-08-10 10:40:56
 */
import { safeNs } from "./utils";

export const book = {
  subscribings: {},

  register(subscribing) {
    let { id } = subscribing;
    this.subscribings[id] = subscribing;
  },

  get(id) {
    return this.subscribings[id];
  },

  remove(id) {
    let subscribing = this.subscribings[id];
    if (subscribing) delete this.subscribings[id];
    return subscribing;
  }
};

export function registerSubscribing(subscribing) {
  let { id } = subscribing;
  SUBSCRIBINGS[id] = subscribing;
}

export const namespaces = {
  default: {}
};

export function getRegistry(ns, create) {
  ns = safeNs(ns);
  let registry = namespaces[ns];
  if (!registry && create) {
    registry = namespaces[ns] = {};
  }
  return registry;
}

export function removeRegistry(ns) {
  delete namespaces[ns];
}


src/subscribe.js
/*
 * @Author: Xavier Yin 
 * @Date: 2018-08-09 17:37:51 
 * @Last Modified by: Xavier Yin
 * @Last Modified time: 2018-08-10 11:09:04
 */

import { safeNs, safeTopic, subId } from "./utils";
import { getRegistry, book } from "./registry";

export default function subscribe(subscriber, topic, callback, options) {
  let { once, ns, ctx } = options || {};
  ns = safeNs(ns);
  topic = safeTopic(topic);
  if (!ctx) ctx = subscriber;
  let registry = getRegistry(ns, true);
  let subscribings = registry[topic] || (registry[topic] = []);
  let id = subId();
  let subscribing = {
    id,
    callback,
    ctx,
    ns,
    once,
    subscriber,
    topic
  };
  book.register(subscribing); // 注册 subscribing
  subscribings.push(subscribing); // 在命名空间中注册 subscribing
  let _armorPubSub =
    subscriber["_armorPubSub"] || (subscriber["_armorPubSub"] = []);
  _armorPubSub.push(subscribing); // 在订阅者中注册 subscribing
  return id;
}


gulpfile.babel.js
import webpack from "webpack";
import gulp from "gulp";
import pkg from "./package.json";
import path from "path";
import WrapperPlugin from "wrapper-webpack-plugin";
import UglifyJSPlugin from "uglifyjs-webpack-plugin";
import Jasmine from "jasmine";

// since webpack 4.0, webpack is built with UglifyJSPlugin if its mode is production.
var webpackConfig = {
  version: pkg.version,

  name: pkg.name,

  entry: "src/index.js",

  build(options) {
    let { mode = "none" } = options || {};
    let isDev = mode !== "production";
    let outputFilename = `${this.name}.js`;
    let outputPath = path.join(__dirname, isDev ? "build" : "dist");

    let plugins = [
      new WrapperPlugin({
        header: `// ${pkg.name} v${
          pkg.version
        } ${new Date().toDateString()}  \n`
      })
    ];
    if (!isDev) {
      plugins.unshift(new UglifyJSPlugin());
    }

    return {
      mode: "none", // it should not set mode into production, because the builtin uglifyjs plugin would remove header from bundle file that is added by WrapperPlugin.
      output: {
        path: outputPath,
        filename: outputFilename,
        library: "ArmorPubSub",
        libraryTarget: "umd",
        globalObject: 'typeof self !== "undefined" ? self : this'
      },
      module: {
        rules: [
          {
            test: /\.js$/,
            exclude: /node_modules/,
            use: {
              loader: "babel-loader",
              query: {
                presets: [["es2015", { loose: true }]]
              }
            }
          }
        ]
      },
      plugins
    };
  },

  release() {
    return this.build({ mode: "production" });
  }
};

function webpackRunner(options) {
  let { release } = options || {};
  return new Promise((resolve, reject) => {
    webpack(webpackConfig[release ? "release" : "build"](), (err, status) => {
      if (err) {
        reject(err);
      } else {
        resolve(status);
      }
    });
  });
}

gulp.task("release", () => webpackRunner({ release: true }));
gulp.task("default", () => webpackRunner());

gulp.task("test", cb => {
  let jasmine = new Jasmine();
  jasmine.loadConfigFile("spec/support/jasmine.json");
  jasmine.onComplete(passed => {
    if (passed) {
      console.log("All specs have passed");
    } else {
      console.log("At least one spec has failed");
    }
    cb();
  });
  jasmine.execute();
});


src/unsubscribe.js
/*
 * @Author: Xavier Yin 
 * @Date: 2018-08-09 18:08:26 
 * @Last Modified by: Xavier Yin
 * @Last Modified time: 2018-08-13 16:12:33
 */

import { queue, isEmptyArray } from "./utils";
import { book, getRegistry, removeRegistry } from "./registry";

function removeSubscribingFromRegistry(subscribing) {
  let { ns, id, topic } = subscribing;
  let registry = getRegistry(ns);
  if (registry) {
    let subscribings = registry[topic];
    if (subscribings) {
      let index = subscribings.findIndex(function(x) {
        return x.id === id;
      });
      subscribings.splice(index, 1);
      if (subscribings.length === 0) {
        delete registry[topic];
        if (Object.keys(registry).length === 0) {
          removeRegistry(ns);
        }
      }
    }
  }
}

function removeSubscribingFromSubscriber(subscribing) {
  let { id, subscriber } = subscribing;
  let pubsub = subscriber._armorPubSub;
  if (pubsub) {
    let index = pubsub.findIndex(function(x) {
      return x.id === id;
    });
    pubsub.splice(index, 1);
  }
}

function unsubscribeApi(subscriber, topic, callback, options) {
  let pubsub = subscriber._armorPubSub;
  if (isEmptyArray(pubsub)) return;
  let { ns, ctx } = options || {};
  let toRemove = [];
  let subscribing, i;
  for (i = 0; i < pubsub.length; i++) {
    subscribing = pubsub[i];
    let { topic: t, callback: cb, ns: n, ctx: c } = subscribing;
    if (
      (topic == void 0 || topic === t) &&
      (callback == void 0 || callback === cb) &&
      (ns == void 0 || ns === n) &&
      (!ctx || ctx === c)
    ) {
      toRemove.push(subscribing);
    }
  }
  for (i = 0; i < toRemove.length; i++) {
    unsubscribeById(toRemove[i]["id"]);
  }
}

export function unsubscribeById(id) {
  let subscribing = book.get(id);
  if (subscribing) {
    book.remove(id);
    removeSubscribingFromRegistry(subscribing);
    removeSubscribingFromSubscriber(subscribing);
  }
}

export function unsubscribeByIdInQueue(id) {
  return queue({ fn: unsubscribeById, args: [id] });
}

export default function unsubscribe(subscriber, topic, callback, options) {
  return queue({
    fn: unsubscribeApi,
    args: [subscriber, topic, callback, options]
  });
}


src/utils.js
/*
 * @Author: Xavier Yin 
 * @Date: 2018-08-09 15:09:07 
 * @Last Modified by: Xavier Yin
 * @Last Modified time: 2018-08-14 10:39:53
 */

/**
 * todo:  暂时不考虑支持以空格分隔多个 Topic 的传参方式
 *
 * 顾虑：
 * 1. 如此，则 topic 被限定为不能包含空格的字符串，在多 topic 的发布便捷性与 topic 字符有效性方面，暂取后者。
 * 2. 多 topic 发布，会导致 sub() 方法返回值类型不行。如果以对象返回（键名为 topic），那会造成 topics 字符串中不能有重复话题。如果以 Array 返回，则对返回值的使用不方便。
 */
//* todo:
// 因为如此的话，则 topic 会被限定为不能包含空格。
// export const EVENT_SPLITTER = /\s+/;
// export const EVENT_TRIM = /(^\s+)|(\s+$)/g;

// export const splitEvent = function(str) {
//   return str.split(EVENT_SPLITTER);
// };

// export const trim = function(str) {
//   return str.replace(EVENT_TRIM, "");
// };

// 强制转换命名空间的合法名称
export function safeNs(ns) {
  return ns == void 0 ? "default" : ns;
}

// 强制转换合法主题
export function safeTopic(topic) {
  return topic == void 0 ? "" : topic;
}

export function isString(obj) {
  return "string" === typeof obj;
}

export function isEmptyArray(arr) {
  return !arr || arr.length === 0;
}

export function isFunction(obj) {
  return typeof obj === "function";
}

// 将 api 调用推入队列中，确保所有 API 顺序调用
export const queue = (function() {
  const _queue = [];
  let isConsuming = false;
  let promise;

  return function(task) {
    _queue.push(task);
    if (isConsuming) return promise;
    let resolve, reject;
    promise = new Promise((rs, rj) => {
      resolve = rs;
      reject = rj;
    });
    isConsuming = true;
    task = _queue.shift();
    try {
      while (task) {
        task.fn(...task.args);
        task = _queue.shift();
      }
      isConsuming = false;
      resolve();
    } catch (e) {
      isConsuming = false;
      reject(e);
    }
    return promise;
  };
})();

// 生成唯一 ID 作为 subscribing 的 ID
export const subId = (function() {
  let count = 0;
  return function() {
    return ++count;
  };
})();
